libp2p/NodeTable.cpp

```diff
@@ -543,8 +543,11 @@ void NodeTable::onPacketReceived(
                     p.sign(m_secret);
                     m_socket->send(p);
 
-                    m_allNodes[in.sourceid]->lastPongSentTime =
-                        RLPXDatagramFace::secondsSinceEpoch();
+                    DEV_GUARDED(x_nodes)
```

Comment:
What's the point of this lock?

Reply:
Accessing `m_allNodes` is protected with this lock everywhere, because it can be acessed from other threads, for example when main thread calls `addNode` when restoring NodeTable state from the file, maybe from some RPC methods, too.

Reply:
This looks like naive protection. You first lock when adding a node, then when updating data. In the middle the entry might have been deleted or modified. It does not look good.

Quick fix would be to check if node exists in the table when updating `lastPongSentTime`. I.e. don't use `[]` operator.

Better might be to set `lastPongSentTime` when adding a node and properly handle the exception of sending PONG (if exception reset `lastPongSentTime`).

Reply:
Well it can be deleted only from private method `dropNode` which is called only in the network thread, so can't be at the same time as this. Here the unsafety is more in other threads (e.g. main thread saving to network.rlp) reading it at the same time while we modify it here.
But I agree it looks not good. We could do similar as in other parts of the class like
```
  if (auto node = nodeEntry(in.sourceid))
    node->lastPongSentTime = RLPXDatagramFace::secondsSinceEpoch();
```
but this actually is also not safe to access `lastPongSentTime` itself this way.

So I think I'll do `find` and modification under one lock as you suggest.